src/lib/services/client-metrics/metrics-service-v2.ts

```diff
@@ -150,12 +155,53 @@ export default class ClientMetricsServiceV2 {
 
     async getClientMetricsForToggle(
         toggleName: string,
-        hoursBack?: number,
+        hoursBack: number = 24,
     ): Promise<IClientMetricsEnv[]> {
-        return this.clientMetricsStoreV2.getMetricsForFeatureToggle(
-            toggleName,
-            hoursBack,
+        const hours = generateHourBuckets(hoursBack);
+        const metrics =
+            await this.clientMetricsStoreV2.getMetricsForFeatureToggle(
+                toggleName,
+                hoursBack,
+            );
+
+        const environments = metrics.reduce(
+            (unique, item) =>
+                unique.includes(item.environment)
+                    ? unique
+                    : [...unique, item.environment],
+            [],
         );
+
+        const applications = metrics.reduce(
+            (unique, item) =>
+                unique.includes(item.appName)
+                    ? unique
+                    : [...unique, item.appName],
+            [],
+        );
+
+        hours.forEach((item: HourBucket) => {
+            if (
+                !metrics.some(
+                    (e) => compareAsc(e.timestamp, item.timestamp) === 0,
+                )
+            ) {
+                environments.forEach((environment) => {
+                    applications.forEach((appName) => {
+                        metrics.push({
```

Comment:
It feels a bit hackish to have 3 for-loops here. Happy to take input on how we can expand the time-series to make sure we have complete "hour buckets" for the requested range (24 hours, 48 hours, etc) for the two supported dimensions (environment and appName).

Reply:
Since the cardinality here is hours * environment * appName, I'm not seeing a way to avoid three for-loops.
Since I'm used to working with streams instead of array I would probably have put the filter first, and then done a concat of the two lists. Feel free to discard the following code suggestion, I just felt like testing if it was possible to do in a pure way. (Well, pure until the concat :) )
```javascript
const fillers = hours.filter(item => !metrics.some(e => compareAsc(e.timestamp, item.timestamp) === 0))
         .flatMap(hourBucket => 
                environments.flatMap(environment => 
                    applications.map(appName => ({ timestamp: item.timestamp, no: 0, yes: 0, appName, environment, featureName: toggleName })
                )
         );
metrics.concat(fillers).sort((a, b) => compareAsc(a.timestamp, b.timestamp))
```

Reply:
Thanks. I think this needs to be changed somehow, because it will not be sufficient to only compare the timestamps, we would need to compare the timestamps, environment and appName, to cover all three dimensions.

Reply:
Hours is 24, environment can't be more than 15, how many appNames are we likely to have? If that's small too then I think this is okay, it looks hackish to have nested loops but if it doesn't actually result in a crazy long loop I think it's okay

Reply:
Thanks for your input, it helped me improve the read-ability somewhat. 

In order to guard against one unknown, number of applications, I have capped it to only consider the first 100 appName's, so the absolute worst case would be 75k items in the array for a feature toggles (`15 envs x 100 appName x 48 hours = 72000`. I think that is fine.